go/store/nbs/cmp_chunk_table_writer.go

```diff
@@ -0,0 +1,321 @@
+// Copyright 2019 Liquidata, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package nbs
+
+import (
+	"crypto/sha512"
+	"encoding/binary"
+	"errors"
+	"hash"
+	"io"
+	"os"
+	"sort"
+
+	"github.com/golang/snappy"
+
+	"github.com/liquidata-inc/dolt/go/libraries/utils/iohelp"
+)
+
+// A ByteSink is an interface for writing bytes which can later be flushed to a writer
+type ByteSink interface {
+	io.Writer
+
+	// Flush writes all the data that was written to the ByteSink to the supplied writer
+	Flush(wr io.Writer) error
+}
+
+// ErrBuffFull used by the FixedBufferSink when the data written is larger than the buffer allocated.
+var ErrBufferFull = errors.New("buffer full")
+
+// FixedBufferByteSink is a ByteSink implementation with a buffer whose size will not change.  Writing more
+// data than the fixed buffer can hold will result in an error
+type FixedBufferByteSink struct {
+	buff []byte
+	pos  uint64
+}
+
+// NewFixedBufferTableSink creates a FixedBufferTableSink which will use the supplied buffer
+func NewFixedBufferTableSink(buff []byte) *FixedBufferByteSink {
+	if len(buff) == 0 {
+		panic("must provide a buffer")
+	}
+
+	return &FixedBufferByteSink{buff: buff}
+}
+
+// Write writes a byte array to the sink.
+func (sink *FixedBufferByteSink) Write(src []byte) (int, error) {
+	dest := sink.buff[sink.pos:]
+	destLen := len(dest)
+	srcLen := len(src)
+
+	if destLen < srcLen {
+		return 0, ErrBufferFull
+	}
+
+	n := copy(dest, src)
+
+	if n != srcLen {
```

Comment:
Is this even possible after the above checks?

Reply:
removed